From b602495b6b385888ba7a2d8aeb3ccabd3920ad79 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Mon, 12 Oct 2026 01:15:49 +0200
Subject: [PATCH] ex5: behaviour library and device options

The example had no options: an options parser now gives the behaviour
library (-l, for instance a library generated by the generic-parallel
interface) and the options of mfem-mgis (declareDefaultOptions), in
particular the device (--device).
---
 ex5/SatohTest.cxx | 24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/ex5/SatohTest.cxx b/ex5/SatohTest.cxx
index 5b1d97e..cdf5801 100644
--- a/ex5/SatohTest.cxx
+++ b/ex5/SatohTest.cxx
@@ -11,6 +11,7 @@
 #include <algorithm>
 #include <fstream>
 #include <iostream>
+#include "mfem/general/optparser.hpp"
 #include "mfem/linalg/vector.hpp"
 #include "mfem/fem/fespace.hpp"
 #include "MFEMMGIS/Parameters.hxx"
@@ -141,8 +142,26 @@ int main(int argc, char** argv) {
   auto or_die = ctx.getFatalFailureHandler();
   //
   static constexpr const auto parallel = false;
-  // options treatment
   mfem_mgis::initialize(argc, argv);
+  // options treatment
+  const char* library = "./src/libBehaviour.so";
+  mfem::OptionsParser args(argc, argv);
+  args.AddOption(&library, "-l", "--library",
+                 "Behaviour library (for instance a library generated by "
+                 "the generic-parallel interface).");
+  // options of mfem-mgis, such as the device (--device)
+  mfem_mgis::declareDefaultOptions(args);
+  args.Parse();
+  if (args.Help()) {
+    args.PrintUsage(mfem_mgis::getOutputStream());
+    mfem_mgis::finalize();
+    return EXIT_SUCCESS;
+  }
+  if (!args.Good()) {
+    args.PrintUsage(mfem_mgis::getOutputStream());
+    mfem_mgis::abort(EXIT_FAILURE);
+  }
+  args.PrintOptions(mfem_mgis::getOutputStream());
   // building the non linear problem
   auto problem =
       mfem_mgis::construct<mfem_mgis::NonLinearEvolutionProblem>(
@@ -158,8 +177,7 @@ int main(int argc, char** argv) {
               {"Parallel", parallel}}) |
       or_die;
   // materials
-  problem.addBehaviourIntegrator(ctx, "Mechanics", "plate",
-                                 "./src/libBehaviour.so",
+  problem.addBehaviourIntegrator(ctx, "Mechanics", "plate", library,
                                  "IsotropicLinearThermoElasticity") |
       or_die;
   auto& m1 = problem.getMaterial(ctx, "plate", 0) | or_die;
-- 
2.43.0

